<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>ClustSIGNAL: a spatial clustering method • clustSIGNAL</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="ClustSIGNAL: a spatial clustering method">
<meta name="description" content="clustSIGNAL: clustering of Spatially Informed Gene expression with Neighbourhood Adapted Learning. A tool for adaptively smoothing and clustering gene expression data. clustSIGNAL uses entropy to measure heterogeneity of cell neighbourhoods and performs a weighted, adaptive smoothing, where homogeneous neighbourhoods are smoothed more and heterogeneous neighbourhoods are smoothed less. This not only overcomes data sparsity but also incorporates spatial context into the gene expression data. The resulting smoothed gene expression data is used for clustering and could be used for other downstream analyses.">
<meta property="og:description" content="clustSIGNAL: clustering of Spatially Informed Gene expression with Neighbourhood Adapted Learning. A tool for adaptively smoothing and clustering gene expression data. clustSIGNAL uses entropy to measure heterogeneity of cell neighbourhoods and performs a weighted, adaptive smoothing, where homogeneous neighbourhoods are smoothed more and heterogeneous neighbourhoods are smoothed less. This not only overcomes data sparsity but also incorporates spatial context into the gene expression data. The resulting smoothed gene expression data is used for clustering and could be used for other downstream analyses.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">clustSIGNAL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.9</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="articles/clustSIGNAL.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sydneybiox/clustSIGNAL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="clustsignal">ClustSIGNAL<a class="anchor" aria-label="anchor" href="#clustsignal"></a>
</h1></div>
<p><img src="reference/figures/clustSIGNAL_hex.jpg" align="right" height="250" width="225"></p>
<p>ClustSIGNAL: <strong><em>Clust</em></strong>ering of <strong><em>S</em></strong>patially <strong><em>I</em></strong>nformed <strong><em>G</em></strong>ene expression with <strong><em>N</em></strong>eighbourhood <strong><em>A</em></strong>dapted <strong><em>L</em></strong>earning.</p>
<p>An R package to perform spatially-informed cell type clustering on high-resolution spatial transcriptomics data. This method has two-fold motivation: overcoming data sparsity by using neighbourhood heterogeneity information to guide an adaptive smoothing approach, and performing spatially-informed clustering by embedding spatial context into the gene expression. To achieve these, we calculate entropy as a measure of heterogeneity of cell neighbourhoods and use it to generate weight distributions and perform adaptive smoothing of gene expression. Homogeneous neighbourhoods generally have low entropy, and so, smoothing is performed over more cells in these neighbourhoods. Contrarily, heterogeneous neighbourhoods have high entropy and are smoothed over a much smaller region. The resulting adaptively smoothed gene expression is used for clustering.</p>
<p>For a tutorial on how to use ClustSIGNAL, see the vignette at this <a href="https://sydneybiox.github.io/clustSIGNAL/">website</a>.</p>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install ClustSIGNAL via Bioconductor:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.github.io/BiocManager/" class="external-link">BiocManager</a></span><span class="op">)</span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"clustSIGNAL"</span><span class="op">)</span></span></code></pre></div>
<p>To install ClustSIGNAL from <a href="https://github.com/SydneyBioX/clustSIGNAL" class="external-link">GitHub</a>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"SydneyBioX/clustSIGNAL"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="method-description">Method description<a class="anchor" aria-label="anchor" href="#method-description"></a>
</h2>
<p><img src="reference/figures/Fig1_schematic.jpg" align="middle"></p>
<p><strong>Figure: ClustSIGNAL method overview.</strong></p>
<p>Here, we present ClustSIGNAL, a spatial clustering method developed to handle data sparsity while considering the variability in cell arrangement of tissue regions. The core steps involved in the method are sequential:</p>
<p><strong>1.</strong> The method starts with non-spatial clustering and subclustering (default louvain clustering) to classify cells into clusters and subclusters that we refer to as ‘initial clusters’ and ‘initial subclusters’, respectively.</p>
<p><strong>2.</strong> The neighbourhood of each cell is then defined in terms of initial subcluster composition.</p>
<p><strong>3.</strong> The cells in the neighbourhood are also sorted and rearranged so that the neighbours belonging to the same initial clusters as the index cell are placed closer to it.</p>
<p><strong>4.</strong> Neighbourhood heterogeneity is measured as entropy, where a high entropy value indicates more heterogeneity in the neighbourhood and a low entropy value indicates a more homogeneous neighbourhood.</p>
<p><strong>5.</strong> The entropy values are used to generate weight distributions specific to each neighbourhood.</p>
<p><strong>6.</strong> The gene expressions of cells are adaptively smoothed using the entropy-guided weight distributions; cells in heterogeneous neighbourhoods (high entropy regions) undergo smoothing over a smaller region, whereas cells in homogeneous neighbourhoods (low entropy regions) undergo smoothing over a larger region.</p>
<p><strong>7.</strong> Non-spatial clustering is performed with adaptively smoothed gene expression to generate ClustSIGNAL clusters.</p>
</div>
<div class="section level2">
<h2 id="clustsignal-parameters">ClustSIGNAL parameters<a class="anchor" aria-label="anchor" href="#clustsignal-parameters"></a>
</h2>
<p>ClustSIGNAL package uses a SpatialExperiment object as input. We provide users with a number of parameters to explore and experiment with, as well as prior tested default values for quick runs. ClustSIGNAL can be used for single sample or multisample analysis with just one function call. Below is the list of the parameters offered and their possible values:</p>
<ul>
<li><p><strong>spe</strong> - SpatialExperiment object with cell spatial coordinates and normalized gene expression in it.</p></li>
<li><p><strong>samples</strong> - column name in cell metadata of the spe object that contains sample names.</p></li>
<li><p><strong>dimRed</strong> - dimensionality reduction method name in reduced dimensions of the spe object. Default value is “None”, in which case PCA is calculated and used as low dimension data.</p></li>
<li><p><strong>batch</strong> - whether batch correction should be performed. Default value is False.</p></li>
<li><p><strong>batch_by</strong> - column name in cell metadata of the spe object containing the groups by which to perform batch correction. By default, batch_by is set to “None”. If the batch option is set to True, the batch_by value must be changed to the group name to perform batch correction.</p></li>
<li><p><strong>NN</strong> - neighbourhood size in terms of the number of nearest neighbours to consider. Value should be &gt; 1. Default value is 30.</p></li>
<li><p><strong>kernel</strong> - type of weight distribution to use. Can be Gaussian (default) or exponential distribution.</p></li>
<li><p><strong>spread</strong> - value of distribution parameter - standard deviation of Gaussian distribution or rate of exponential distribution. Default value is 0.3, recommended for Gaussian distribution. For exponential distribution, recommended value is 5.</p></li>
<li><p><strong>sort</strong> - whether cell neighbourhoods should be sorted by their initial cluster labels. Default value is True.</p></li>
<li><p><strong>threads</strong> - number of cpus to use for parallel runs. Default value is 1.</p></li>
<li><p><strong>outputs</strong> - choice of output types. Default value is ‘c’ for a list containing data frame of cell IDs and ClustSIGNAL cluster labels. Other possible values are “n” for a list of ClustSIGNAL clusters dataframe and neighbourhood matrix, “s” for a list of ClustSIGNAL clusters dataframe and final SpatialExperiment object, and “a” for a list of all 3 outputs - ClustSIGNAL clusters dataframe, neighbourhood matrix, and final spe object.</p></li>
<li>
<p><strong>clustParams</strong> - parameter options for TwoStepParam clustering methods in the bluster package. The clustering parameters are in the following order in this list:</p>
<ul>
<li><p><strong><em>clust_c</em></strong> represents number of centers for clustering with KmeansParam. By default this is set to 0, in which case the method uses either 5000 centers or 1/5th of the total cells in the data as the number of centers, whichever is lower.</p></li>
<li><p><strong><em>subclust_c</em></strong> represents number of centers for sub-clustering the initial clusters with KmeansParam. The default value is 0, in which case the method uses either 1 center or half of the total cells in the initial cluster as the number of centers, whichever is higher.</p></li>
<li><p><strong><em>iter.max</em></strong> for the maximum number of iterations to perform during clustering and subclustering with KmeansParam. Default value is 30.</p></li>
<li><p><strong><em>k</em></strong> indicates the k-value used for clustering/subclustering with NNGraphParam. Default value is 10.</p></li>
<li><p><strong><em>cluster.fun</em></strong> is the graph clustering method used with NNGraphParam. By default, the Louvain method is used.</p></li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="running-clustsignal">Running ClustSIGNAL<a class="anchor" aria-label="anchor" href="#running-clustsignal"></a>
</h2>
<p>Before running ClustSIGNAL, it is important to ensure that the input SpatialExperiment object has spatial coordinates stored in the spatialCoords matrix. Otherwise, the method will throw an error asking the user to provide spatial coordinates.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/clustSIGNAL/">clustSIGNAL</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">ClustSignal_example</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Here, the sample labels are in 'sample_id' column.</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="reference/clustSIGNAL.html">clustSIGNAL</a></span><span class="op">(</span><span class="va">spe</span>, samples <span class="op">=</span> <span class="st">"sample_id"</span>, outputs <span class="op">=</span> <span class="st">"a"</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/sydneybiox/clustSIGNAL/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/sydneybiox/clustSIGNAL/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-2" class="external-link">GPL-2</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing clustSIGNAL</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Pratibha Panwar <br><small class="roles"> Maintainer, author, contributor </small> <a href="https://orcid.org/0000-0002-7437-7084" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Boyi Guo <br><small class="roles"> Author </small>  </li>
<li>Haowen Zhao <br><small class="roles"> Author </small>  </li>
<li>Stephanie Hicks <br><small class="roles"> Author </small>  </li>
<li>Shila Ghazanfar <br><small class="roles"> Author, contributor </small> <a href="https://orcid.org/0000-0001-7861-6997" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pratibha Panwar, Boyi Guo, Haowen Zhao, Stephanie Hicks, Shila Ghazanfar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
