<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>ClustSIGNAL — clustSIGNAL • clustSIGNAL</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ClustSIGNAL — clustSIGNAL"><meta name="description" content="A clustering method for cell type classification of spatial transcriptomics
data. The tool generates and uses an adaptively smoothed, spatially informed
gene expression data for clustering."><meta property="og:description" content="A clustering method for cell type classification of spatial transcriptomics
data. The tool generates and uses an adaptively smoothed, spatially informed
gene expression data for clustering."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">clustSIGNAL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.6</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/clustSIGNAL.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sydneybiox/clustSIGNAL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>ClustSIGNAL</h1>
      <small class="dont-index">Source: <a href="https://github.com/sydneybiox/clustSIGNAL/blob/main/R/clustSIGNAL.R" class="external-link"><code>R/clustSIGNAL.R</code></a></small>
      <div class="d-none name"><code>clustSIGNAL.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>A clustering method for cell type classification of spatial transcriptomics
data. The tool generates and uses an adaptively smoothed, spatially informed
gene expression data for clustering.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">clustSIGNAL</span><span class="op">(</span></span>
<span>  <span class="va">spe</span>,</span>
<span>  <span class="va">samples</span>,</span>
<span>  dimRed <span class="op">=</span> <span class="st">"None"</span>,</span>
<span>  batch <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  batch_by <span class="op">=</span> <span class="st">"None"</span>,</span>
<span>  NN <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  kernel <span class="op">=</span> <span class="st">"G"</span>,</span>
<span>  spread <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  sort <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  threads <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  outputs <span class="op">=</span> <span class="st">"c"</span>,</span>
<span>  clustParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>clust_c <span class="op">=</span> <span class="fl">0</span>, subclust_c <span class="op">=</span> <span class="fl">0</span>, iter.max <span class="op">=</span> <span class="fl">30</span>, k <span class="op">=</span> <span class="fl">5</span>, cluster.fun <span class="op">=</span></span>
<span>    <span class="st">"louvain"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-spe">spe<a class="anchor" aria-label="anchor" href="#arg-spe"></a></dt>
<dd><p>a SpatialExperiment object.</p></dd>


<dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>a character indicating name of colData(spe) column containing
sample names.</p></dd>


<dt id="arg-dimred">dimRed<a class="anchor" aria-label="anchor" href="#arg-dimred"></a></dt>
<dd><p>a character indicating the name of the reduced dimensions to
use from the SpatialExperiment object (i.e., from reducedDimNames(spe)).
Default value is 'None'.</p></dd>


<dt id="arg-batch">batch<a class="anchor" aria-label="anchor" href="#arg-batch"></a></dt>
<dd><p>a logical parameter for whether or not to perform batch
correction. Default value is FALSE.</p></dd>


<dt id="arg-batch-by">batch_by<a class="anchor" aria-label="anchor" href="#arg-batch-by"></a></dt>
<dd><p>a character indicating name of colData(spe) column containing
the batch names.</p></dd>


<dt id="arg-nn">NN<a class="anchor" aria-label="anchor" href="#arg-nn"></a></dt>
<dd><p>an integer for the number of neighbourhood cells the function
should consider. The value must be greater than or equal to 1. Default value
is 30.</p></dd>


<dt id="arg-kernel">kernel<a class="anchor" aria-label="anchor" href="#arg-kernel"></a></dt>
<dd><p>a character for type of distribution to be used. The two valid
values are "G" or "E". G for Gaussian distribution, and E for exponential
distribution. Default value is "G".</p></dd>


<dt id="arg-spread">spread<a class="anchor" aria-label="anchor" href="#arg-spread"></a></dt>
<dd><p>a numeric value for distribution spread, represented by
standard deviation for Gaussian distribution and rate for exponential
distribution. Default value is 0.05 for Gaussian distribution and 20 for
exponential distribution.</p></dd>


<dt id="arg-sort">sort<a class="anchor" aria-label="anchor" href="#arg-sort"></a></dt>
<dd><p>a logical parameter for whether or not to sort the neighbourhood
after region description. Default value is TRUE.</p></dd>


<dt id="arg-threads">threads<a class="anchor" aria-label="anchor" href="#arg-threads"></a></dt>
<dd><p>a numeric value for the number of CPU cores to be used for the
analysis. Default value set to 1.</p></dd>


<dt id="arg-outputs">outputs<a class="anchor" aria-label="anchor" href="#arg-outputs"></a></dt>
<dd><p>a character for the type of output to return to the user. "c"
for data frame of cell IDs and their respective cluster numbers (default),
"n" for dataframe of clusters plus neighbourhood matrix, "s" for  dataframe
of clusters plus final spatialExperiment object, or "a" for all outputs.</p></dd>


<dt id="arg-clustparams">clustParams<a class="anchor" aria-label="anchor" href="#arg-clustparams"></a></dt>
<dd><p>a list of parameters for TwoStepParam clustering methods.
The clustering parameters are in the order - centers (centers) for clustering
with KmeansParam, centers (centers) for sub-clustering clusters with
KmeansParam, maximum iterations (iter.max) for clustering with KmeansParam,
k values (k) for clustering with NNGraphParam, and community detection method
(cluster.fun) to use with NNGraphParam.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>a list of outputs</p>
<p>1. clusters: a data frame of cell names and their cluster classification.</p>
<p>2. neighbours: a character matrix containing cells IDs of each cell's
neighbours</p>
<p>3. spe_final: a SpatialExperiment object with initial 'putative cell type'
groups, entropy values, smoothed gene expression, post-smoothing clusters,
and silhouette widths included.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">example</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">spe</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "uniqueID"    "sample_id"   "entropy"     "nsCluster"   "initCluster"</span>
<span class="r-in"><span><span class="co"># identify the column names with cell and sample labels</span></span></span>
<span class="r-in"><span><span class="va">samples</span> <span class="op">=</span> <span class="st">"sample_id"</span></span></span>
<span class="r-in"><span><span class="va">res_list</span> <span class="op">&lt;-</span> <span class="fu">clustSIGNAL</span><span class="op">(</span><span class="va">spe</span>, <span class="va">samples</span>, outputs <span class="op">=</span> <span class="st">"c"</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Calculating PCA. Time 05:14:50"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ClustSIGNAL run started. Time 05:14:50"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Initial nonspatial clustering performed. Clusters = 6 Time 05:14:50"</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You're computing too large a percentage of total singular values, use a standard svd instead.</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>You're computing too large a percentage of total singular values, use a standard svd instead.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Nonspatial subclustering performed. Subclusters = 17 Time 05:14:52"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Regions defined. Time 05:14:52"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Region domainness calculated. Time 05:14:52"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Smoothing performed. NN = 30 Kernel = G Spread = 0.05 Time 05:14:52"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "Nonspatial clustering performed on smoothed data. Clusters = 9 Time 05:14:52"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "ClustSIGNAL run completed. 05:14:52"</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Time difference of 2.5301 secs</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pratibha Panwar, Boyi Guo, Haowen Zhao, Stephanie Hicks, Shila Ghazanfar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

