<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>ClustSIGNAL tutorial • clustSIGNAL</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ClustSIGNAL tutorial">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">clustSIGNAL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.8</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/clustSIGNAL.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/sydneybiox/clustSIGNAL/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>ClustSIGNAL tutorial</h1>
                        <h4 data-toc-skip class="author">Pratibha
Panwar, Boyi Guo, Haowen Zhou, Stephanie Hicks, Shila Ghazanfar</h4>
            
            <h4 data-toc-skip class="date">2025-03-21</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sydneybiox/clustSIGNAL/blob/main/vignettes/clustSIGNAL.Rmd" class="external-link"><code>vignettes/clustSIGNAL.Rmd</code></a></small>
      <div class="d-none name"><code>clustSIGNAL.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>In this vignette, we demonstrate how spatially-resolved clustering
can be performed with ClustSIGNAL, exploring the clusters using
pre-defined metrics like adjusted rand index (ARI) and normalized mutual
information (NMI), as well as spatial plots to visualize the clusters.
ClustSIGNAL is a multisample spatial clustering approach, and we show
this using an example dataset. We also display the use of entropy
measures, which are generated as a by-product of the ClustSIGNAL
process, in understanding the tissue structure of a sample.</p>
<p>ClustSIGNAL is very flexible in that it allows for, (i) user-provided
input values for most parameters (default parameter values are also
provided) and (ii) running ClustSIGNAL step-by-step. This tutorial
demonstrates how the step-by-step clustering can be performed, and what
parameters need to be defined at each step.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load required packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sydneybiox.github.io/clustSIGNAL/">clustSIGNAL</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/jchiquet/aricode" class="external-link">aricode</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="single-sample-analysis-with-clustsignal">Single sample analysis with ClustSIGNAL<a class="anchor" aria-label="anchor" href="#single-sample-analysis-with-clustsignal"></a>
</h2>
<p>In this section, we use the SeqFISH mouse embryo dataset from <a href="https://www.nature.com/articles/s41587-021-01006-2" class="external-link">Lohoff et al,
2021</a>, which contains spatial transcriptomics data from 3 mouse
embryos, with 351 genes and 57,536 cells. For this vignette, we have
subset the data by randomly selecting 5000 cells from Embryo 2,
excluding cells that had been manually annotated as ‘Low quality’.</p>
<p>We begin by creating a SpatialExperiment object from the gene
expression and cell information in the data subset, ensuring that the
spatial coordinates are stored in spatialCoords within the
SpatialExperiment object. If the data are already in a SpatialExperiment
object, ClustSIGNAL can be run as long as basic requirements like
spatial coordinates, normalized counts, and unique cell names are
met.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load me_expr containing gene expression logcounts</span></span>
<span><span class="co"># load me_data containing cell metadata including x-y coordinates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mEmbryo2</span><span class="op">)</span></span>
<span><span class="co"># to create a SpatialExperiment object we need gene expression, cell metadata, </span></span>
<span><span class="co"># and cell locations.</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">SpatialExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment.html" class="external-link">SpatialExperiment</a></span><span class="op">(</span></span>
<span>  assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logcounts <span class="op">=</span> <span class="va">me_expr</span><span class="op">)</span>, </span>
<span>  colData <span class="op">=</span> <span class="va">me_data</span>,</span>
<span>  <span class="co"># spatialCoordsNames requires column names in me_data that contain </span></span>
<span>  <span class="co"># xy-coordinates of cells</span></span>
<span>  spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">spe</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 351 5000 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): logcounts</span></span>
<span><span class="co">## rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(5000): embryo2_Pos29_cell100_z2 embryo2_Pos29_cell101_z5 ...</span></span>
<span><span class="co">##   embryo2_Pos50_cell97_z5 embryo2_Pos50_cell99_z5</span></span>
<span><span class="co">## colData names(4): uniqueID pos celltype_mapped_refined sample_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(2) : X Y</span></span>
<span><span class="co">## imgData names(0):</span></span></code></pre>
<p>For running ClustSIGNAL, we need to know the column name in colData
slot of the SpatialExperiment object that contains the sample labels.
Here, the sample labels are in the ‘sample_id’ column.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># column names in the metadata</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "uniqueID"                "pos"                    </span></span>
<span><span class="co">## [3] "celltype_mapped_refined" "sample_id"</span></span></code></pre>
<div class="section level3">
<h3 id="running-clustsignal-on-one-sample">Running ClustSIGNAL on one sample<a class="anchor" aria-label="anchor" href="#running-clustsignal-on-one-sample"></a>
</h3>
<p>The simplest ClustSIGNAL run requires a SpatialExperiment object, the
colData column name of sample labels, and the type of output to
generate.</p>
<p>Other parameters that can be modified include: (i) dimRed - specifies
the low dimension data to use (default ‘None’); (ii) batch - when TRUE,
ClustSIGNAL performs batch correction and needs a valid value for
batch_by; (iii) batch_by - name of metadata column containing sample
batches contributing to batch effect (default ‘None’); (iv) NN -
specifies the neighbourhood size (default 30); (v) kernel - specifies
the distribution to use for weight generation (default ‘G’ for
Gaussian); (vi) spread - the distribution spread value (default 0.3 for
Gaussian); (vii) sort - when TRUE, ClustSIGNAL sorts the neighbourhood;
(viii) threads - specifies the number of cpus to use in parallel runs
(default 1); and (ix) clustParams - list of parameters to use for
non-spatial clustering components.</p>
<p>Furthermore, the adaptively smoothed gene expression data generated
by ClustSIGNAL could be useful for other downstream analyses and is
accessible if the output options ‘s’ or ‘a’ are selected to return the
final SpatialExperiment object.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="st">"sample_id"</span> <span class="co"># column name containing sample names</span></span>
<span><span class="co"># to run ClustSIGNAL, requires a SpatialExperiment object, column name of sample</span></span>
<span><span class="co"># labels in colData slot, and the output type to generate (clusters, neighbours,</span></span>
<span><span class="co"># and/or final spe object).</span></span>
<span><span class="va">res_emb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustSIGNAL.html">clustSIGNAL</a></span><span class="op">(</span><span class="va">spe</span>, <span class="va">samples</span>, outputs <span class="op">=</span> <span class="st">"a"</span><span class="op">)</span> </span></code></pre></div>
<pre><code><span><span class="co">## [1] "Calculating PCA. Time 06:35:36"</span></span>
<span><span class="co">## [1] "ClustSIGNAL run started. Time 06:35:37"</span></span>
<span><span class="co">## [1] "Initial nonspatial clustering performed. Clusters = 9 Time 06:35:38"</span></span>
<span><span class="co">## [1] "Nonspatial subclustering performed. Subclusters = 31 Time 06:35:40"</span></span>
<span><span class="co">## [1] "Regions defined. Time 06:35:42"</span></span>
<span><span class="co">## [1] "Region heterogeneity calculated. Time 06:35:42"</span></span>
<span><span class="co">## [1] "Smoothing performed. NN = 30 Kernel = G Spread = 0.3 Time 06:35:42"</span></span>
<span><span class="co">## [1] "Nonspatial clustering performed on smoothed data. Clusters = 14 Time 06:35:44"</span></span>
<span><span class="co">## [1] "ClustSIGNAL run completed. 06:35:44"</span></span>
<span><span class="co">## Time difference of 7.587263 secs</span></span></code></pre>
<p>This returns a list that contains a ClustSIGNAL clusters dataframe
(clusters), a matrix of cell IDs from each cell’s neighbourhood
(neighbours with NN neighbourhood size), and a final SpatialExperiment
object (spe_final).</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_emb</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># names of the outputs generated</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "clusters"   "neighbours" "spe_final"</span></span></code></pre>
<p>The cluster dataframe contains cell IDs and their cluster labels
assigned by ClustSIGNAL.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">res_emb</span><span class="op">$</span><span class="va">clusters</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># cluster data frame has cell IDs and cluster labels</span></span></code></pre></div>
<pre><code><span><span class="co">##                      Cells Clusters</span></span>
<span><span class="co">## 1 embryo2_Pos29_cell100_z2       12</span></span>
<span><span class="co">## 2 embryo2_Pos29_cell101_z5       12</span></span>
<span><span class="co">## 3 embryo2_Pos29_cell104_z2       12</span></span>
<span><span class="co">## 4 embryo2_Pos29_cell104_z5       12</span></span>
<span><span class="co">## 5 embryo2_Pos29_cell105_z2       12</span></span>
<span><span class="co">## 6 embryo2_Pos29_cell108_z5       12</span></span></code></pre>
<p>The output SpatialExperiment object contains the adaptively smoothed
gene expression data as an additional assay (smoothed), as well as
initial clusters and subclusters, entropy values, and ClustSIGNAL
clusters.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for convenience with downstream analyses, we will replace the original spe</span></span>
<span><span class="co"># object with the one generated by ClustSIGNAL. This does not lead to any loss </span></span>
<span><span class="co"># of information as ClustSIGNAL only adds information to the input spe object.</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">res_emb</span><span class="op">$</span><span class="va">spe_final</span></span>
<span><span class="va">spe</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 351 5000 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): logcounts smoothed</span></span>
<span><span class="co">## rownames(351): Abcc4 Acp5 ... Zfp57 Zic3</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(5000): embryo2_Pos29_cell100_z2 embryo2_Pos29_cell101_z5 ...</span></span>
<span><span class="co">##   embryo2_Pos50_cell97_z5 embryo2_Pos50_cell99_z5</span></span>
<span><span class="co">## colData names(8): uniqueID pos ... entropy ClustSIGNAL</span></span>
<span><span class="co">## reducedDimNames(2): PCA PCA.smooth</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(2) : X Y</span></span>
<span><span class="co">## imgData names(1): sample_id</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">|&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "uniqueID"                "pos"                    </span></span>
<span><span class="co">## [3] "celltype_mapped_refined" "sample_id"              </span></span>
<span><span class="co">## [5] "initCluster"             "initSubcluster"         </span></span>
<span><span class="co">## [7] "entropy"                 "ClustSIGNAL"</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="visualising-clustsignal-clusters">Visualising ClustSIGNAL clusters<a class="anchor" aria-label="anchor" href="#visualising-clustsignal-clusters"></a>
</h3>
<p>We use spatial coordinates of cells and their ClustSIGNAL cluster
labels and entropy values to visualize the clustering output.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#635547"</span>, <span class="st">"#8EC792"</span>, <span class="st">"#9e6762"</span>, <span class="st">"#FACB12"</span>, <span class="st">"#3F84AA"</span>, <span class="st">"#0F4A9C"</span>, </span>
<span>            <span class="st">"#ff891c"</span>, <span class="st">"#EF5A9D"</span>, <span class="st">"#C594BF"</span>, <span class="st">"#DFCDE4"</span>, <span class="st">"#139992"</span>, <span class="st">"#65A83E"</span>, </span>
<span>            <span class="st">"#8DB5CE"</span>, <span class="st">"#005579"</span>, <span class="st">"#C9EBFB"</span>, <span class="st">"#B51D8D"</span>, <span class="st">"#532C8A"</span>, <span class="st">"#8870ad"</span>, </span>
<span>            <span class="st">"#cc7818"</span>, <span class="st">"#FBBE92"</span>, <span class="st">"#EF4E22"</span>, <span class="st">"#f9decf"</span>, <span class="st">"#c9a997"</span>, <span class="st">"#C72228"</span>, </span>
<span>            <span class="st">"#f79083"</span>, <span class="st">"#F397C0"</span>, <span class="st">"#DABE99"</span>, <span class="st">"#c19f70"</span>, <span class="st">"#354E23"</span>, <span class="st">"#C3C388"</span>,</span>
<span>            <span class="st">"#647a4f"</span>, <span class="st">"#CDE088"</span>, <span class="st">"#f7f79e"</span>, <span class="st">"#F6BFCB"</span>, <span class="st">"#7F6874"</span>, <span class="st">"#989898"</span>, </span>
<span>            <span class="st">"#1A1A1A"</span>, <span class="st">"#FFFFFF"</span>, <span class="st">"#e6e6e6"</span>, <span class="st">"#77441B"</span>, <span class="st">"#F90026"</span>, <span class="st">"#A10037"</span>, </span>
<span>            <span class="st">"#DA5921"</span>, <span class="st">"#E1C239"</span>, <span class="st">"#9DD84A"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for plotting with scater R package, we need to add the spatial coordinates </span></span>
<span><span class="co"># to the reduced dimension slot of the spe object</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">spe</span>, <span class="st">"spatial"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># spatial plot</span></span>
<span><span class="va">spt_clust</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span></span>
<span>  <span class="va">spe</span>, colour_by <span class="op">=</span> <span class="st">"ClustSIGNAL"</span>, dimred <span class="op">=</span> <span class="st">"spatial"</span>, point_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  point_size <span class="op">=</span> <span class="fl">4</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"A. Spatial plot of clusters"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Clusters"</span>, </span>
<span>                               override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Scale for <span style="color: #00BB00;">colour</span> is already present.</span></span>
<span><span class="co">## Adding another scale for <span style="color: #00BB00;">colour</span>, which will replace the existing scale.</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># entropy distribution plotted at cluster-level can indicate which clusters </span></span>
<span><span class="co"># have cells from homogeneous/heterogeneous space. </span></span>
<span><span class="va">df_met</span> <span class="op">&lt;-</span> <span class="va">spe</span> <span class="op">|&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">ct_ent</span> <span class="op">&lt;-</span> <span class="va">df_met</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>ClustSIGNAL <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">ClustSIGNAL</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">ClustSIGNAL</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="co"># calculating median entropy of each cluster category</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mdEntropy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">entropy</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="co"># reordering clusters by their median entropy value</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">mdEntropy</span><span class="op">)</span></span>
<span><span class="va">df_met</span><span class="op">$</span><span class="va">ClustSIGNAL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">df_met</span><span class="op">$</span><span class="va">ClustSIGNAL</span>, levels <span class="op">=</span> <span class="va">ct_ent</span><span class="op">$</span><span class="va">ClustSIGNAL</span><span class="op">)</span></span>
<span><span class="va">col_ent</span> <span class="op">&lt;-</span> <span class="va">colors</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">ct_ent</span><span class="op">$</span><span class="va">ClustSIGNAL</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">box_clust</span> <span class="op">&lt;-</span> <span class="va">df_met</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ClustSIGNAL</span>, y <span class="op">=</span> <span class="va">entropy</span>, fill <span class="op">=</span> <span class="va">ClustSIGNAL</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_ent</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"B. Entropy distribution of clusters"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ClustSIGNAL clusters"</span>, y <span class="op">=</span> <span class="st">"Entropy"</span>, name <span class="op">=</span> <span class="st">"Clusters"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>        text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span>, hjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>        plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spt_clust</span> <span class="op">+</span> <span class="va">box_clust</span> <span class="op">+</span> <span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span>, </span>
<span>                                               widths <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustSIGNAL_files/figure-html/embryo_spatialPlots4-1.png" width="700"></p>
<p>The spatial location and entropy distribution of the clusters provide
spatial context of the cells and their neighbourhoods, as well as the
compositions of the neighbourhoods. For example, in panel (B) the low
entropy clusters are generally found in space that is more homogeneous,
whereas the high entropy clusters belong to neighbourhoods that have
more cell diversity. This can also be visualized in the spatial plot in
panel (A).</p>
</div>
<div class="section level3">
<h3 id="assessing-clustering-accuracy">Assessing clustering accuracy<a class="anchor" aria-label="anchor" href="#assessing-clustering-accuracy"></a>
</h3>
<p>We assess the clustering efficiency of ClustSIGNAL using the commonly
used clustering metrics ARI and NMI, which are usable only when prior
cell annotations are available. Here, ARI and NMI measure the similarity
or agreement (respectively) between cluster labels obtained from
ClustSIGNAL and manual cell annotations.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># to assess the accuracy of clustering, the cluster labels are often compared to</span></span>
<span><span class="co"># prior annotations. Here, we compare ClustSIGNAL cluster labels to annotations </span></span>
<span><span class="co"># available with this public data.</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span></span>
<span>    ARI <span class="op">=</span> <span class="fu">aricode</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/aricode/man/ARI.html" class="external-link">ARI</a></span><span class="op">(</span><span class="va">celltype_mapped_refined</span>, <span class="va">ClustSIGNAL</span><span class="op">)</span>, <span class="co"># calculate ARI</span></span>
<span>    NMI <span class="op">=</span> <span class="fu">aricode</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/aricode/man/NMI.html" class="external-link">NMI</a></span><span class="op">(</span><span class="va">celltype_mapped_refined</span>, <span class="va">ClustSIGNAL</span><span class="op">)</span><span class="op">)</span> <span class="co"># calculate NMI</span></span></code></pre></div>
<pre><code><span><span class="co">##         ARI       NMI</span></span>
<span><span class="co">## 1 0.3117711 0.5663197</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="entropy-spread-and-distribution">Entropy spread and distribution<a class="anchor" aria-label="anchor" href="#entropy-spread-and-distribution"></a>
</h3>
<p>The entropy values generated through ClustSIGNAL process can be
useful in analyzing the sample structure.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we can assess the overall entropy distribution of the dataset</span></span>
<span><span class="va">spe</span> <span class="op">|&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>min_Entropy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">entropy</span><span class="op">)</span>,</span>
<span>            min_Entropy_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">entropy</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>            max_Entropy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">entropy</span><span class="op">)</span>,</span>
<span>            mean_Entropy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">entropy</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   min_Entropy min_Entropy_count max_Entropy mean_Entropy</span></span>
<span><span class="co">## 1           0               382      3.1518     1.119852</span></span></code></pre>
<p>The entropy range can indicate whether the tissue sample contains any
homogeneous regions. For example, a min_Entropy of 0 means that some
cells are placed in completely homogeneous space when looking at a
neighbourhood size of 30 cells (NN = 30 was used for generating the
entropy values). The min_Entropy_count gives us an idea of the total
number of such low entropy neighbourhoods in the sample.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># we can also visualize the distribution and spread of the entropy values</span></span>
<span><span class="va">hst_ent</span> <span class="op">&lt;-</span> <span class="va">spe</span> <span class="op">|&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">entropy</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"A. Entropy spread"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Entropy"</span>, y <span class="op">=</span> <span class="st">"Number of neighbourhoods"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>        plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spt_ent</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe</span>, colour_by <span class="op">=</span> <span class="st">"entropy"</span>,</span>
<span>                                    <span class="co"># specify spatial low dimension</span></span>
<span>                                    dimred <span class="op">=</span> <span class="st">"spatial"</span>, point_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                    point_size <span class="op">=</span> <span class="fl">4</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ggtitle</a></span><span class="op">(</span><span class="st">"B. Entropy spatial distribution"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient2</a></span><span class="op">(</span><span class="st">"Entropy"</span>, low <span class="op">=</span> <span class="st">"grey"</span>, high <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html" class="external-link">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">spe</span><span class="op">$</span><span class="va">entropy</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Scale for <span style="color: #00BB00;">colour</span> is already present.</span></span>
<span><span class="co">## Adding another scale for <span style="color: #00BB00;">colour</span>, which will replace the existing scale.</span></span></code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hst_ent</span> <span class="op">+</span> <span class="va">spt_ent</span></span></code></pre></div>
<p><img src="clustSIGNAL_files/figure-html/entropyPlots3-1.png" width="700"></p>
<p>The spread and spatial distribution of neighbourhood entropies can be
useful in visually assessing and comparing tissue compositions in
samples - low entropy neighbourhoods are more homogeneous and likely
contain cell type-specific niches, whereas high entropy neighbourhoods
are heterogeneous with more uniform distribution of different cell
types.</p>
</div>
</div>
<div class="section level2">
<h2 id="multisample-analysis-with-clustsignal">Multisample analysis with ClustSIGNAL<a class="anchor" aria-label="anchor" href="#multisample-analysis-with-clustsignal"></a>
</h2>
<p>Here, we use the MERFISH mouse hypothalamus preoptic region dataset
from <a href="https://www.science.org/doi/10.1126/science.aau5324" class="external-link">Moffitt et
al, 2018</a>, which contains spatial transcriptomics data from 181
samples, with 155 genes and 1,027,080 cells. For this vignette, we have
subset the data by selecting 6000 random cells from only 3 samples -
Animal 1 Bregma -0.09 (2080 cells), Animal 7 Bregma 0.16 (1936 cells),
and Animal 7 Bregma -0.09 (1984 cells), excluding cells that were
manually annotated as ‘Ambiguous’ and 20 genes for which expression was
generated using a different technology.</p>
<p>We start the analysis by creating a SpatialExperiment object from the
gene expression and cell information in the data subset, ensuring that
the spatial coordinates are stored in spatialCoords slot within the spe
object.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load mh_expr containing gene expression logcounts</span></span>
<span><span class="co"># load mh_data containing cell metadata and cell x-y coordinates</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mHypothal</span><span class="op">)</span></span>
<span><span class="co"># create spe object using gene expression, cell metadata, and cell locations</span></span>
<span><span class="va">spe2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment.html" class="external-link">SpatialExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logcounts <span class="op">=</span> <span class="va">mh_expr</span><span class="op">)</span>, </span>
<span>                          colData <span class="op">=</span> <span class="va">mh_data</span>,</span>
<span>                          <span class="co"># spatialCoordsNames requires column names in </span></span>
<span>                          <span class="co"># mh_data that contain xy-coordinates of cells</span></span>
<span>                          spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">spe2</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 135 6000 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): logcounts</span></span>
<span><span class="co">## rownames(135): Ace2 Adora2a ... Ttn Ttyh2</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(6000): 74d3f69d-e8f2-4c33-a8ca-fac3eb65e55a</span></span>
<span><span class="co">##   41158ddc-e70c-487b-b891-0cb3c8452555 ...</span></span>
<span><span class="co">##   54145623-7071-482c-b9da-d0d2dd31274a</span></span>
<span><span class="co">##   96bc85ce-b993-4fb1-8e0c-165f83f0cfd0</span></span>
<span><span class="co">## colData names(4): Cell_ID Cell_class sample_id samples</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(2) : X Y</span></span>
<span><span class="co">## imgData names(0):</span></span></code></pre>
<p>Next we identify sample labels column in the SpatialExperiment
object.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe2</span> <span class="op">|&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># metadata summary</span></span></code></pre></div>
<pre><code><span><span class="co">## Formal class 'DFrame' [package "S4Vectors"] with 6 slots</span></span>
<span><span class="co">##   ..@ rownames       : chr [1:6000] "74d3f69d-e8f2-4c33-a8ca-fac3eb65e55a" "41158ddc-e70c-487b-b891-0cb3c8452555" "46ba8016-2c4f-4ef0-84c9-3ee3951afdfd" "ac9f1af8-8b03-4b2f-b29c-929ae2b240dc" ...</span></span>
<span><span class="co">##   ..@ nrows          : int 6000</span></span>
<span><span class="co">##   ..@ elementType    : chr "ANY"</span></span>
<span><span class="co">##   ..@ elementMetadata: NULL</span></span>
<span><span class="co">##   ..@ metadata       : list()</span></span>
<span><span class="co">##   ..@ listData       :List of 4</span></span>
<span><span class="co">##   .. ..$ Cell_ID   : chr [1:6000] "74d3f69d-e8f2-4c33-a8ca-fac3eb65e55a" "41158ddc-e70c-487b-b891-0cb3c8452555" "46ba8016-2c4f-4ef0-84c9-3ee3951afdfd" "ac9f1af8-8b03-4b2f-b29c-929ae2b240dc" ...</span></span>
<span><span class="co">##   .. ..$ Cell_class: chr [1:6000] "Endothelial 1" "Endothelial 1" "Pericytes" "Astrocyte" ...</span></span>
<span><span class="co">##   .. ..$ sample_id : chr [1:6000] "sample01" "sample01" "sample01" "sample01" ...</span></span>
<span><span class="co">##   .. ..$ samples   : Factor w/ 3 levels "1.-0.09","7.-0.09",..: 1 1 1 1 1 1 1 1 1 1 ...</span></span></code></pre>
<p>Here, the sample labels are in the ‘samples’ column of the
object.</p>
<div class="section level3">
<h3 id="clustsignal-run">ClustSIGNAL run<a class="anchor" aria-label="anchor" href="#clustsignal-run"></a>
</h3>
<p>An important concept to take into account when running multisample
analysis is batch effects. When gathering samples from different sources
or through different technologies/procedures, some technical batch
effects might be introduced into the dataset. We can run ClustSIGNAL in
batch correction mode simply by setting batch = TRUE and batch_by =
“group”, where group will be the name of the colData column of spe
object that contains the batch information. ClustSIGNAL then uses <a href="https://portals.broadinstitute.org/harmony/" class="external-link">harmony</a>
internally for batch correction.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">110</span><span class="op">)</span></span>
<span><span class="co"># ClustSIGNAL can be run on a dataset with multiple samples. As before, we need</span></span>
<span><span class="co"># the SpatialExperiment object and column name of sample labels in the object. </span></span>
<span><span class="co"># The method can be run in parallel through the threads option. Here we use </span></span>
<span><span class="co"># thread = 4 to use 4 cores.</span></span>
<span><span class="co"># Since no batch effects were observed in this data subset, we have not used </span></span>
<span><span class="co"># the batch and batch_by options.</span></span>
<span><span class="va">samples</span> <span class="op">&lt;-</span> <span class="st">"samples"</span> <span class="co"># column name containing sample names</span></span>
<span><span class="va">res_hyp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/clustSIGNAL.html">clustSIGNAL</a></span><span class="op">(</span><span class="va">spe2</span>, <span class="va">samples</span>, threads <span class="op">=</span> <span class="fl">4</span>, outputs <span class="op">=</span> <span class="st">"a"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Calculating PCA. Time 06:35:47"</span></span>
<span><span class="co">## [1] "ClustSIGNAL run started. Time 06:35:47"</span></span>
<span><span class="co">## [1] "Initial nonspatial clustering performed. Clusters = 9 Time 06:35:48"</span></span>
<span><span class="co">## [1] "Nonspatial subclustering performed. Subclusters = 38 Time 06:35:50"</span></span>
<span><span class="co">## [1] "Regions defined. Time 06:35:53"</span></span>
<span><span class="co">## [1] "Region heterogeneity calculated. Time 06:35:54"</span></span>
<span><span class="co">## [1] "Smoothing performed. NN = 30 Kernel = G Spread = 0.3 Time 06:35:54"</span></span>
<span><span class="co">## [1] "Nonspatial clustering performed on smoothed data. Clusters = 12 Time 06:35:55"</span></span>
<span><span class="co">## [1] "ClustSIGNAL run completed. 06:35:56"</span></span>
<span><span class="co">## Time difference of 8.33502 secs</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for convenience with downstream analyses, we replace the original spe object </span></span>
<span><span class="co"># with the one generated by ClustSIGNAL.</span></span>
<span><span class="va">spe2</span> <span class="op">&lt;-</span> <span class="va">res_hyp</span><span class="op">$</span><span class="va">spe_final</span></span>
<span><span class="va">spe2</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 135 6000 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(2): logcounts smoothed</span></span>
<span><span class="co">## rownames(135): Ace2 Adora2a ... Ttn Ttyh2</span></span>
<span><span class="co">## rowData names(0):</span></span>
<span><span class="co">## colnames(6000): 74d3f69d-e8f2-4c33-a8ca-fac3eb65e55a</span></span>
<span><span class="co">##   41158ddc-e70c-487b-b891-0cb3c8452555 ...</span></span>
<span><span class="co">##   54145623-7071-482c-b9da-d0d2dd31274a</span></span>
<span><span class="co">##   96bc85ce-b993-4fb1-8e0c-165f83f0cfd0</span></span>
<span><span class="co">## colData names(8): Cell_ID Cell_class ... entropy ClustSIGNAL</span></span>
<span><span class="co">## reducedDimNames(2): PCA PCA.smooth</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(2) : X Y</span></span>
<span><span class="co">## imgData names(1): sample_id</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="clustering-metrics">Clustering metrics<a class="anchor" aria-label="anchor" href="#clustering-metrics"></a>
</h3>
<p>Clustering and entropy results can be calculated and visualized for
each sample.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">samplesList</span> <span class="op">&lt;-</span> <span class="va">spe2</span><span class="op">[[</span><span class="va">samples</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># get sample names</span></span>
<span><span class="va">samplesList</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "1.-0.09" "7.-0.09" "7.0.16"</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe2</span> <span class="op">|&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span></span>
<span>    <span class="co"># Comparing ClustSIGNAL cluster labels to annotations available with the </span></span>
<span>    <span class="co"># public data to assess its accuracy.</span></span>
<span>    ARI <span class="op">=</span> <span class="fu">aricode</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/aricode/man/ARI.html" class="external-link">ARI</a></span><span class="op">(</span><span class="va">Cell_class</span>, <span class="va">ClustSIGNAL</span><span class="op">)</span>,</span>
<span>    NMI <span class="op">=</span> <span class="fu">aricode</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/aricode/man/NMI.html" class="external-link">NMI</a></span><span class="op">(</span><span class="va">Cell_class</span>, <span class="va">ClustSIGNAL</span><span class="op">)</span>,</span>
<span>    <span class="co"># Assessing the overall entropy distribution of the samples in the dataset.</span></span>
<span>    min_Entropy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">entropy</span><span class="op">)</span>,</span>
<span>    min_Entropy_count <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">entropy</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>    max_Entropy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">entropy</span><span class="op">)</span>,</span>
<span>    mean_Entropy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">entropy</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 3 × 7</span></span></span>
<span><span class="co">##   samples   ARI   NMI min_Entropy min_Entropy_count max_Entropy mean_Entropy</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> 1.-0.09 0.381 0.513       1.19                  0        4.17         3.19</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> 7.-0.09 0.398 0.555       1.08                  0        4.24         3.15</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> 7.0.16  0.587 0.613       0.883                 0        4.31         3.08</span></span></code></pre>
<p>As before, the entropy range can tell us a lot about the tissue
structure of the samples. Unlike the seqFISH subset data, where the
minimum entropy of the sample was 0, here, the minimum entropy is higher
indicating that the tissue doesn’t really have any cell type-specific
niches when looking at neighbourhood size of 30 cells. Moreover, the
relatively high mean entropy value indicates that the tissues slices are
quite heterogeneous.</p>
</div>
<div class="section level3">
<h3 id="visualizing-clustsignal-clusters">Visualizing ClustSIGNAL clusters<a class="anchor" aria-label="anchor" href="#visualizing-clustsignal-clusters"></a>
</h3>
<p>ClustSIGNAL performs clustering on all cells in the dataset in one
run, thereby generating the same clusters across multiple samples. The
cluster labels do not need to be mapped between samples. For example,
cluster 1 represents the same cell type in all three samples, without
needing explicit mapping between samples.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># for plotting with scater R package, we need to add the spatial coordinates </span></span>
<span><span class="co"># to the reduced dimension section</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/reducedDims.html" class="external-link">reducedDim</a></span><span class="op">(</span><span class="va">spe2</span>, <span class="st">"spatial"</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment-methods.html" class="external-link">spatialCoords</a></span><span class="op">(</span><span class="va">spe2</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># spatial plot - ClustSIGNAL clusters</span></span>
<span><span class="va">spt_clust2</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe2</span>, colour_by <span class="op">=</span> <span class="st">"ClustSIGNAL"</span>,</span>
<span>                                    <span class="co"># specify spatial low dimension</span></span>
<span>                                    dimred <span class="op">=</span> <span class="st">"spatial"</span>, point_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                    point_size <span class="op">=</span> <span class="fl">4</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">spe2</span><span class="op">[[</span><span class="va">samples</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free"</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Clusters"</span>,</span>
<span>                               override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Scale for <span style="color: #00BB00;">colour</span> is already present.</span></span>
<span><span class="co">## Adding another scale for <span style="color: #00BB00;">colour</span>, which will replace the existing scale.</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># For visualising cluster-level entropy distribution, we reorder the clusters </span></span>
<span><span class="co"># by their median entropy value in each sample</span></span>
<span><span class="va">df_met2</span> <span class="op">&lt;-</span> <span class="va">spe2</span> <span class="op">|&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">box_clust2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">s</span> <span class="kw">in</span> <span class="va">samplesList</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">df_met_sub</span> <span class="op">&lt;-</span> <span class="va">df_met2</span><span class="op">[</span><span class="va">df_met2</span><span class="op">[[</span><span class="va">samples</span><span class="op">]</span><span class="op">]</span> <span class="op">==</span> <span class="va">s</span>, <span class="op">]</span></span>
<span>  <span class="co"># calculating median entropy of each cluster in a sample</span></span>
<span>  <span class="va">ct_ent2</span> <span class="op">&lt;-</span> <span class="va">df_met_sub</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>ClustSIGNAL <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">ClustSIGNAL</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">ClustSIGNAL</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarise</a></span><span class="op">(</span>mdEntropy <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="va">entropy</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="co"># reordering clusters by their median entropy</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">mdEntropy</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="va">df_met_sub</span><span class="op">$</span><span class="va">ClustSIGNAL</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">df_met_sub</span><span class="op">$</span><span class="va">ClustSIGNAL</span>, </span>
<span>                                   levels <span class="op">=</span> <span class="va">ct_ent2</span><span class="op">$</span><span class="va">ClustSIGNAL</span><span class="op">)</span></span>
<span>  <span class="co"># box plot of cluster entropy</span></span>
<span>  <span class="va">col_ent2</span> <span class="op">&lt;-</span> <span class="va">colors</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="va">ct_ent2</span><span class="op">$</span><span class="va">ClustSIGNAL</span><span class="op">)</span><span class="op">]</span></span>
<span>  <span class="va">box_clust2</span><span class="op">[[</span><span class="va">s</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">df_met_sub</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">ClustSIGNAL</span>, y <span class="op">=</span> <span class="va">entropy</span>, fill <span class="op">=</span> <span class="va">ClustSIGNAL</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">col_ent2</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"ClustSIGNAL clusters"</span>, y <span class="op">=</span> <span class="st">"Entropy"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">ylim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>strip.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>          legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>          text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,</span>
<span>          axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spt_clust2</span> <span class="op">/</span> <span class="op">(</span><span class="fu">patchwork</span><span class="fu">::</span><span class="fu"><a href="https://patchwork.data-imaginist.com/reference/wrap_plots.html" class="external-link">wrap_plots</a></span><span class="op">(</span><span class="va">box_clust2</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>                <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>axes <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span>, heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">3</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span></span>
<span>    title <span class="op">=</span> <span class="st">"Spatial (top) and entropy (bottom) distributions of clusters"</span>,</span>
<span>    theme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustSIGNAL_files/figure-html/hypothal_spatialPlots4-1.png" width="700"></p>
<p>The spatial location and entropy distribution of the clusters can be
compared in a multisample analysis, providing spatial context of the
cluster cells and their neighbourhood compositions in the different
samples within the dataset. Since the clusters were generated in a
single run, they are same across the different samples. Therefore, if a
cluster is not represented in a sample, this would mean that its
respective cell type is not present in that sample.</p>
</div>
<div class="section level3">
<h3 id="visualising-entropy-spread-and-distribution">Visualising entropy spread and distribution<a class="anchor" aria-label="anchor" href="#visualising-entropy-spread-and-distribution"></a>
</h3>
<p>In multisample analysis, tissue structure of the different samples in
the dataset can be compared using the spread and spatial distribution of
the neighbourhood entropy measures.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hst_ent2</span> <span class="op">&lt;-</span> <span class="va">spe2</span> <span class="op">|&gt;</span> <span class="fu">colData</span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">entropy</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>binwidth <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">samples</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Entropy"</span>, y <span class="op">=</span> <span class="st">"Number of neighbourhoods"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spt_ent2</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">spe2</span>, colour_by <span class="op">=</span> <span class="st">"entropy"</span>,</span>
<span>                                  <span class="co"># specify spatial low dimension</span></span>
<span>                                  dimred <span class="op">=</span> <span class="st">"spatial"</span>, point_alpha <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                  point_size <span class="op">=</span> <span class="fl">4</span>, scattermore <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_gradient.html" class="external-link">scale_colour_gradient2</a></span><span class="op">(</span><span class="st">"Entropy"</span>, low <span class="op">=</span> <span class="st">"grey"</span>, high <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html" class="external-link">scale_size_continuous</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">spe2</span><span class="op">$</span><span class="va">entropy</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">spe2</span><span class="op">[[</span><span class="va">samples</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free"</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>strip.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>        text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Scale for <span style="color: #00BB00;">colour</span> is already present.</span></span>
<span><span class="co">## Adding another scale for <span style="color: #00BB00;">colour</span>, which will replace the existing scale.</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hst_ent2</span> <span class="op">/</span> <span class="va">spt_ent2</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>heights <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span></span>
<span>      title <span class="op">=</span> <span class="st">"Entropy spread (top) and spatial distribution (bottom)"</span>,</span>
<span>      theme <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="clustSIGNAL_files/figure-html/hypothal_entropyPlots3-1.png" width="700"></p>
<p>Together, these plots help in visually assessing tissue compositions
of the samples - all 3 samples have high entropy neighbourhoods
indicating that they mainly have heterogeneous regions with uniform
distribution of different cell types.</p>
</div>
</div>
<div class="section level2">
<h2 id="clustsignal-step-by-step-run">ClustSIGNAL step-by-step run<a class="anchor" aria-label="anchor" href="#clustsignal-step-by-step-run"></a>
</h2>
<p>ClustSIGNAL has five main functions for each distinct step in its
algorithm. These functions are accessible and can be run sequentially to
generate data from intermediate steps, if needed. For example,
ClustSIGNAL can be run step-by-step up to the entropy measurement
component, without having to run the complete method. The entropy values
will be added to the SpatialExperiment object and can be used for
assessing tissue structure in terms of its heterogeneity. Similarly, the
adaptively smoothed gene expression can be obtained by running
ClustSIGNAL till the adaptive smoothing step. Here, we describe how
individual ClustSIGNAL functions can be used sequentially.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load logcounts and metadata to the environment</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">mEmbryo2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># as before, we read the data into a SpatialExperiment object</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/SpatialExperiment.html" class="external-link">SpatialExperiment</a></span><span class="op">(</span>assays <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>logcounts <span class="op">=</span> <span class="va">me_expr</span><span class="op">)</span>,</span>
<span>                         colData <span class="op">=</span> <span class="va">me_data</span>, spatialCoordsNames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="co"># first we need to generate low dimension data for initial clustering</span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">scater</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocSingular/man/runPCA.html" class="external-link">runPCA</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span> </span></code></pre></div>
<div class="section level3">
<h3 id="step-1-initial-clustering-and-subclustering">Step 1: Initial clustering and subclustering<a class="anchor" aria-label="anchor" href="#step-1-initial-clustering-and-subclustering"></a>
</h3>
<p>The first step in the ClustSIGNAL algorithm is initial clustering and
subclustering. For this, we need to provide a spe object with low
embedding information.</p>
<p>Other parameters have default values: batch = FALSE and batch_by =
“None” (if no batch correction needs to be performed), threads = 1,
clustParams = list(clust_c = 0, subclust_c = 0, iter.max = 30, k = 10,
cluster.fun = “louvain”).</p>
<p>Among the clustering parameters, clust_c and subclust_c refer to the
number of centers to use for clustering and sub-clustering with
KmeansParam. By default clust_c is set to 0, in which case the method
uses either 5000 centers or 1/5th of the total cells in the data as the
number of centers, whichever is lower. Similarly, subclust_c is set to 0
by default, in which case the method uses either 1 center or half of the
total cells in the initial cluster as the number of centers, whichever
is higher. For all other values of clust_c and subclust_c, the input is
treated as the number of centers.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">clustSIGNAL</span><span class="fu">::</span><span class="fu"><a href="../reference/p1_clustering.html">p1_clustering</a></span><span class="op">(</span><span class="va">spe</span>, dimRed <span class="op">=</span> <span class="st">"PCA"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Initial nonspatial clustering performed. Clusters = 9 Time 06:36:01"</span></span>
<span><span class="co">## [1] "Nonspatial subclustering performed. Subclusters = 31 Time 06:36:03"</span></span></code></pre>
<p>Here, two columns are added to the spe object under the cell
metadata:</p>
<ol style="list-style-type: lower-roman">
<li>the initial cluster labels,</li>
</ol>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span><span class="op">$</span><span class="va">initCluster</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># clustering output</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7 7 7 7 3 7</span></span>
<span><span class="co">## Levels: 1 2 3 4 5 6 7 8 9</span></span></code></pre>
<ol start="2" style="list-style-type: lower-roman">
<li>the initial subcluster labels.</li>
</ol>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span><span class="op">$</span><span class="va">initSubcluster</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># subclustering output</span></span></code></pre></div>
<pre><code><span><span class="co">## embryo2_Pos29_cell100_z2 embryo2_Pos29_cell101_z5 embryo2_Pos29_cell104_z2 </span></span>
<span><span class="co">##                      7.1                      7.1                      7.1 </span></span>
<span><span class="co">## embryo2_Pos29_cell104_z5 embryo2_Pos29_cell105_z2 embryo2_Pos29_cell108_z5 </span></span>
<span><span class="co">##                      7.1                      3.3                      7.1 </span></span>
<span><span class="co">## 31 Levels: 1.1 1.2 1.3 2.1 2.2 2.3 2.4 3.1 3.2 3.3 4.1 4.2 4.3 4.4 5.1 ... 9.4</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="step-2-neighbourhood-detection">Step 2: Neighbourhood detection<a class="anchor" aria-label="anchor" href="#step-2-neighbourhood-detection"></a>
</h3>
<p>The next step involves detecting the neighborhood of all cells. We
need the spe object containing the initial cluster and initial
subcluster labels and sample IDs for this.</p>
<p>By default, ClustSIGNAL identifies 30 nearest neighbors (NN = 30),
sorts the neighbourhood (sort = TRUE), and does not use parallel runs
(threads = 1).</p>
<p>ClustSIGNAL allows the use of external cell labels generated through
other methods, in place of the initial clusters and subclusters. For
this, the cell cluster and subcluster labels of each cell must be stored
in the colData of the spe object as “initCluster” and “initSubcluster”,
respectively.</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This step generates a list of neighbourhood information.</span></span>
<span><span class="va">outReg</span> <span class="op">&lt;-</span> <span class="fu">clustSIGNAL</span><span class="fu">::</span><span class="fu"><a href="../reference/neighbourDetect.html">neighbourDetect</a></span><span class="op">(</span><span class="va">spe</span>, samples <span class="op">=</span> <span class="st">"sample_id"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Regions defined. Time 06:36:05"</span></span></code></pre>
<p>This generates a list containing:</p>
<ol style="list-style-type: lower-roman">
<li>a neighborhood matrix containing cell IDs,</li>
</ol>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outReg</span><span class="op">$</span><span class="va">nnCells</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">##                          [,1]                       [,2]                      </span></span>
<span><span class="co">## embryo2_Pos29_cell100_z2 "embryo2_Pos29_cell100_z2" "embryo2_Pos29_cell90_z5" </span></span>
<span><span class="co">## embryo2_Pos29_cell101_z5 "embryo2_Pos29_cell101_z5" "embryo2_Pos29_cell117_z2"</span></span>
<span><span class="co">## embryo2_Pos29_cell104_z2 "embryo2_Pos29_cell104_z2" "embryo2_Pos29_cell94_z5" </span></span>
<span><span class="co">##                          [,3]                      </span></span>
<span><span class="co">## embryo2_Pos29_cell100_z2 "embryo2_Pos29_cell104_z2"</span></span>
<span><span class="co">## embryo2_Pos29_cell101_z5 "embryo2_Pos29_cell97_z5" </span></span>
<span><span class="co">## embryo2_Pos29_cell104_z2 "embryo2_Pos29_cell100_z2"</span></span></code></pre>
<ol start="2" style="list-style-type: lower-roman">
<li>a list of arrays containing initial subcluster proportions.</li>
</ol>
<div class="sourceCode" id="cb63"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outReg</span><span class="op">$</span><span class="va">regXclust</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span> </span></code></pre></div>
<pre><code><span><span class="co">## arr</span></span>
<span><span class="co">##        3.3        6.1        6.2        7.1        7.2        9.2 </span></span>
<span><span class="co">## 0.03225806 0.03225806 0.03225806 0.83870968 0.03225806 0.03225806</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="step-3-entropy-measure">Step 3: Entropy measure<a class="anchor" aria-label="anchor" href="#step-3-entropy-measure"></a>
</h3>
<p>Now that we know the neighbourhood of each cell, we can calculate
entropy of each cell’s neighborhood. For this, we need the spe object
and initial subcluster proportions. This step can run in parallel, but
by default we use 1 cpu core.</p>
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">clustSIGNAL</span><span class="fu">::</span><span class="fu"><a href="../reference/entropyMeasure.html">entropyMeasure</a></span><span class="op">(</span><span class="va">spe</span>, <span class="va">outReg</span><span class="op">$</span><span class="va">regXclust</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Region heterogeneity calculated. Time 06:36:06"</span></span></code></pre>
<p>The entropy values are added to the spe object under cell
metadata.</p>
<div class="sourceCode" id="cb67"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span><span class="op">$</span><span class="va">entropy</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># entropy values</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 1.01189 0.20559 0.61207 0.54755 0.20559 0.00000</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="step-4-adaptive-smoothing">Step 4: Adaptive smoothing<a class="anchor" aria-label="anchor" href="#step-4-adaptive-smoothing"></a>
</h3>
<p>Using the entropy values, we can perform adaptive smoothing. This
requires the spe object containing the entropy values as well as the
neighborhood matrix of cell IDs generated during neighbourhood
detection.</p>
<p>Other parameters for which default values are provided include number
of neighbors (NN = 30), weight distribution type (kernel = “G” for
Gaussian), distribution spread (spread = 0.05 representing standard
deviation for Gaussian distribution; for exponential distribution we
recommend using a spread of 5 indicating rate of the distribution), and
number of cores (threads = 1) to use for parallel runs.</p>
<div class="sourceCode" id="cb69"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">clustSIGNAL</span><span class="fu">::</span><span class="fu"><a href="../reference/adaptiveSmoothing.html">adaptiveSmoothing</a></span><span class="op">(</span><span class="va">spe</span>, <span class="va">outReg</span><span class="op">$</span><span class="va">nnCells</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Smoothing performed. NN = 30 Kernel = G Spread = 0.3 Time 06:36:06"</span></span></code></pre>
<p>The adaptively smoothed gene expression data are added to the spe
object under assays as ‘smoothed’.</p>
<div class="sourceCode" id="cb71"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">assay</span><span class="op">(</span><span class="va">spe</span>, <span class="st">"smoothed"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## 5 x 3 sparse Matrix of class "dgCMatrix"</span></span>
<span><span class="co">##         embryo2_Pos29_cell100_z2 embryo2_Pos29_cell101_z5</span></span>
<span><span class="co">## Abcc4                 0.32885379                0.2841961</span></span>
<span><span class="co">## Acp5                  0.08959616                0.1189281</span></span>
<span><span class="co">## Acvr1                 0.40086591                0.3359335</span></span>
<span><span class="co">## Acvr2a                0.51895151                0.5066584</span></span>
<span><span class="co">## Adora2b               0.12246298                0.1257018</span></span>
<span><span class="co">##         embryo2_Pos29_cell104_z2</span></span>
<span><span class="co">## Abcc4                 0.32878008</span></span>
<span><span class="co">## Acp5                  0.09205565</span></span>
<span><span class="co">## Acvr1                 0.39842503</span></span>
<span><span class="co">## Acvr2a                0.47449863</span></span>
<span><span class="co">## Adora2b               0.08577944</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="step-5-final-clustering">Step 5: Final clustering<a class="anchor" aria-label="anchor" href="#step-5-final-clustering"></a>
</h3>
<p>The final step involves performing clustering on the adaptively
smoothed data. We only need to provide the spe object containing the
adaptively smoothed data. This step has the same default clustering and
batch correction parameters as the initial clustering in first step.</p>
<div class="sourceCode" id="cb73"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">clustSIGNAL</span><span class="fu">::</span><span class="fu"><a href="../reference/p2_clustering.html">p2_clustering</a></span><span class="op">(</span><span class="va">spe</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Nonspatial clustering performed on smoothed data. Clusters = 14 Time 06:36:07"</span></span></code></pre>
<p>Cluster labels are added to the colData of the spe object under a
ClustSIGNAL column</p>
<div class="sourceCode" id="cb75"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe</span><span class="op">$</span><span class="va">ClustSIGNAL</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># ClustSIGNAL cluster labels</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 12 12 12 12 12 12</span></span>
<span><span class="co">## Levels: 1 2 3 4 5 6 7 8 9 10 11 12 13 14</span></span></code></pre>
<details><summary><strong>Session Information</strong>
</summary><div class="sourceCode" id="cb77"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.3 (2025-02-28)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">## Running under: Ubuntu 24.04.2 LTS</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">##  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">## [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: system (glibc)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] aricode_1.0.3               patchwork_1.3.0            </span></span>
<span><span class="co">##  [3] dplyr_1.1.4                 scater_1.34.1              </span></span>
<span><span class="co">##  [5] ggplot2_3.5.1               scuttle_1.16.0             </span></span>
<span><span class="co">##  [7] clustSIGNAL_0.99.8          SpatialExperiment_1.16.0   </span></span>
<span><span class="co">##  [9] SingleCellExperiment_1.28.1 SummarizedExperiment_1.36.0</span></span>
<span><span class="co">## [11] Biobase_2.66.0              GenomicRanges_1.58.0       </span></span>
<span><span class="co">## [13] GenomeInfoDb_1.42.3         IRanges_2.40.1             </span></span>
<span><span class="co">## [15] S4Vectors_0.44.0            BiocGenerics_0.52.0        </span></span>
<span><span class="co">## [17] MatrixGenerics_1.18.1       matrixStats_1.5.0          </span></span>
<span><span class="co">## [19] BiocStyle_2.34.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] gridExtra_2.3           rlang_1.1.5             magrittr_2.0.3         </span></span>
<span><span class="co">##  [4] compiler_4.4.3          systemfonts_1.2.1       vctrs_0.6.5            </span></span>
<span><span class="co">##  [7] reshape2_1.4.4          stringr_1.5.1           pkgconfig_2.0.3        </span></span>
<span><span class="co">## [10] crayon_1.5.3            fastmap_1.2.0           magick_2.8.5           </span></span>
<span><span class="co">## [13] XVector_0.46.0          labeling_0.4.3          utf8_1.2.4             </span></span>
<span><span class="co">## [16] rmarkdown_2.29          UCSC.utils_1.2.0        ggbeeswarm_0.7.2       </span></span>
<span><span class="co">## [19] ragg_1.3.3              xfun_0.51               bluster_1.16.0         </span></span>
<span><span class="co">## [22] zlibbioc_1.52.0         cachem_1.1.0            beachmat_2.22.0        </span></span>
<span><span class="co">## [25] jsonlite_1.9.1          DelayedArray_0.32.0     BiocParallel_1.40.0    </span></span>
<span><span class="co">## [28] irlba_2.3.5.1           parallel_4.4.3          cluster_2.1.8          </span></span>
<span><span class="co">## [31] R6_2.6.1                bslib_0.9.0             stringi_1.8.4          </span></span>
<span><span class="co">## [34] scattermore_1.2         jquerylib_0.1.4         Rcpp_1.0.14            </span></span>
<span><span class="co">## [37] bookdown_0.42           knitr_1.50              Matrix_1.7-2           </span></span>
<span><span class="co">## [40] igraph_2.1.4            tidyselect_1.2.1        abind_1.4-8            </span></span>
<span><span class="co">## [43] yaml_2.3.10             viridis_0.6.5           codetools_0.2-20       </span></span>
<span><span class="co">## [46] lattice_0.22-6          tibble_3.2.1            plyr_1.8.9             </span></span>
<span><span class="co">## [49] withr_3.0.2             evaluate_1.0.3          desc_1.4.3             </span></span>
<span><span class="co">## [52] pillar_1.10.1           BiocManager_1.30.25     generics_0.1.3         </span></span>
<span><span class="co">## [55] munsell_0.5.1           scales_1.3.0            glue_1.8.0             </span></span>
<span><span class="co">## [58] tools_4.4.3             BiocNeighbors_2.0.1     ScaledMatrix_1.14.0    </span></span>
<span><span class="co">## [61] fs_1.6.5                cowplot_1.1.3           grid_4.4.3             </span></span>
<span><span class="co">## [64] colorspace_2.1-1        GenomeInfoDbData_1.2.13 beeswarm_0.4.0         </span></span>
<span><span class="co">## [67] BiocSingular_1.22.0     vipor_0.4.7             cli_3.6.4              </span></span>
<span><span class="co">## [70] rsvd_1.0.5              textshaping_1.0.0       S4Arrays_1.6.0         </span></span>
<span><span class="co">## [73] viridisLite_0.4.2       gtable_0.3.6            sass_0.4.9             </span></span>
<span><span class="co">## [76] digest_0.6.37           SparseArray_1.6.2       ggrepel_0.9.6          </span></span>
<span><span class="co">## [79] rjson_0.2.23            farver_2.1.2            htmltools_0.5.8.1      </span></span>
<span><span class="co">## [82] pkgdown_2.1.1           lifecycle_1.0.4         httr_1.4.7             </span></span>
<span><span class="co">## [85] harmony_1.2.3</span></span></code></pre>
</details>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pratibha Panwar, Boyi Guo, Haowen Zhao, Stephanie Hicks, Shila Ghazanfar.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
